File: javascript/examples/catalog-parser.ts
import {
  BaseParser,
  ItemRetry,
  configField,
  defineConfig,
  type ConfigObject,
  type ParseSet,
  type RequestOptions,
} from 'stream-miner-sdk';

interface CatalogConfig extends ConfigObject {
  start_url: string;
  retries: number;
}

export default class CatalogParser extends BaseParser<CatalogConfig> {
  static parserAlias = 'catalog_example_js';

  static Config = defineConfig({
    start_url: configField('https://example.test/catalog.json', {
      label: 'Start URL',
      description: 'Catalog endpoint used when the queue item is empty.',
    }),
    retries: configField(2, {
      label: 'Retries',
      type: 'number',
      description: 'Per-request retry count for catalog fetches.',
    }),
  });

  async parse(set: ParseSet, results: Record<string, unknown>) {
    const url = String(set.query || this.conf.start_url);
    const options: RequestOptions = {
      retries: this.conf.retries,
      allowStatuses: [200],
      validateResponse: response => response.ok,
    };

    const response = await this.get(url, {}, options);
    if (!response.ok) {
      throw new ItemRetry('catalog_http_error', {
        maxAttempts: 3,
        retryDelaySec: 30,
        metadata: {status: response.status, url},
      });
    }

    const lease = await this.egress.next({
      httpMode: this.runtime.http_transport_mode,
      reason: 'catalog_fetch',
    });
    const payload = response.json();
    const outputPath = 'catalog/last-response.json';

    await this.metric('catalog_pages_seen', 1, {
      transport: String(lease.transport_mode || 'unknown'),
    });
    await this.publishJson('catalog-last-response', payload, {
      metadata: {url},
    });
    await this.writeText(outputPath, JSON.stringify(payload, null, 2));
    await this.registerOutput('last-response', outputPath, {
      outputKind: 'file',
      contentType: 'application/json',
      metadata: {url},
    });

    results.url = url;
    results.dataset = 'catalog-last-response';
    return results;
  }
}

File: python/examples/catalog_parser.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from stream_miner_sdk import BaseParser, EgressSelector, ItemRetry, RequestOptions, config_field


@dataclass
class CatalogConfig:
    start_url: str = config_field(
        "https://example.test/catalog.json",
        label="Start URL",
        description="Catalog endpoint used when the queue item is empty.",
    )
    retries: int = config_field(
        2,
        label="Retries",
        type="number",
        description="Per-request retry count for catalog fetches.",
    )


class CatalogParser(BaseParser[CatalogConfig]):
    __parser_alias__ = "catalog_example_py"
    Config = CatalogConfig

    async def parse(self, query: Any) -> None:
        url = str(query or self.conf.start_url)
        response = await self.get(
            url,
            options=RequestOptions(
                retries=self.conf.retries,
                allow_statuses={200},
                validate_response=lambda item: item.ok,
            ),
        )
        if not response.ok:
            raise ItemRetry(
                "catalog_http_error",
                max_attempts=3,
                retry_delay_sec=30,
                metadata={"status": response.status, "url": url},
            )

        lease = await self.egress.next(
            EgressSelector(http_mode=self.runtime.http_transport_mode),
            reason="catalog_fetch",
        )
        payload = response.json()
        output_path = await self.write_text("catalog/last-response.json", response.text())

        await self.metric(
            "catalog_pages_seen",
            1,
            {"transport": str(lease.transport_mode or "unknown")},
        )
        await self.publish_json(
            name="catalog-last-response",
            payload=payload,
            metadata={"url": url},
        )
        await self.register_output(
            handle="last-response",
            path=output_path,
            output_kind="file",
            content_type="application/json",
            metadata={"url": url},
        )
